_posts/2020-10-23-deconvolution.md

```diff
@@ -0,0 +1,446 @@
+---
+layout: post
+title: Image Analysis Redux
+tagline: Dask + CuPy + RL
+author: John Kirkham (NVIDIA) and Ben Zaitlen (NVIDIA)
+theme: twitter
+---
+{% include JB/setup %}
+
+Summary
+-------
+
+[Last year](https://blog.dask.org/2019/08/09/image-itk) we experimented with
+Dask/ITK/Scikit-Image to perform large scale image analysis on a stack of 3D
+images.  Specifically, we looked at `deconvolution`, a common method to
+_deblur_ images.  Now, a year later, we return to these experiments with a
+better understanding of how Dask and CuPy can interact, enhanced serialization
+methods, and support from the open-source community.
+
+Image Analysis Redux
+--------------------
+
+Previously we used the [Richardson Lucy
+(RL)](https://en.wikipedia.org/wiki/Richardson%E2%80%93Lucy_deconvolution)
+deconvolution algorithm from ITK and
+[Scikit-Image](https://github.com/scikit-image/scikit-image/blob/master/skimage/restoration/deconvolution.py#L329).
+We left off at theorizing how GPUs could potentially help accelerate these
+workflows.  Starting with Scikit-Image's implementation, we naively tried
+replacing `scipy.signal.convolve` calls with `cupyx.scipy.ndimage.convolve`,
+and while performance improved, it did not improve _significantly_ -- that is,
+we did not get the 100X speed we were looking for.
+
+As it often turns out in mathematics a problem that can be inefficient to solve
+in one representation can often be made more efficent by transforming the data
+beforehand. In this new representation we can solve the same problem
+(convolution in this case) more easily before transforming the result back into
+a more familiar representation. When it comes to convolution, the
+transformation we apply is called [Fast-Fourirer Transform
+(FFT)](https://en.wikipedia.org/wiki/Fast_Fourier_transform). Once this is
+applied we are able to convolve data using a simple multiplication.
+
+As it turns out this FFT transformation is extremely fast on both CPUs and
+GPUs. Similarly the algorithm we can write with FFTs are accelerated.  Despite
+the added step of doing FFTs, the cost of transformation + the cost of the
+algorithm is still lower than performing the original algorithm in real space.
+We (and others before us) found this was the case for Richardson Lucy (on both
+CPUs and GPUs) and performance continued increasing when we parallelized with
+Dask over multiple GPUs.
+
+Help from Open-Source
+---------------------
+
+An FFT RL equivalent has been around for some time and the good folks at the
+[Solar Dynamics Observatory](https://sdo.gsfc.nasa.gov/mission/instruments.php)
+built and shared a NumPy/CuPy implementation as part the [Atmospheric Imaging
+Assembly](https://aiapy.readthedocs.io/en/v0.2.0/_modules/aiapy/psf/deconvolve.html)
+Python package (aiapy).  We slightly modified their implementation to handle 3D
```

Comment:
I cannot follow this. You mention that you support 3-dimensional point spread functions in the text, but in the deconvolutoin code you just select one slice if the PSF has dimensions other than 2? And then there is padding in x and y dimensions. So are you trying to do 2d deconvolution on individual slices of the volume? In particular for the 3D microscopy dataset further down I would assume that the PSF is indeed 3 dimensional?

Reply:
Yeah the code probably needs to be updated. Initially we had a PSF that had an extra singleton dimension for some reason that we removed. We did some cleanup of the code, but we may have missed some things still. Thanks for pointing that out!

Reply:
Ooooh, random extra singleton dimensions make a lot more sense. Definitely be good to clean that up before this goes out!

Reply:
Ok I think we've addressed this issue. Wound up doing a bit of rewriting of this function. The resulting code is a bit simpler and avoids branches around dimensions. Hopefully this makes more sense. Please let us know if anything else sticks out here 🙂